### PR TITLE
Add Metric Reader and use IEnumerable for Metric collect instead of Batch

### PR DESCRIPTION
**Pull request by cijothomas**
1. Trying the to-be-merged spec for MetricReader : https://github.com/open-telemetry/opentelemetry-specification/pull/1888/files
Note that we don't support multiple MetricReaders currently, this will be added in follow ups.

2. Replace Batch<Metric> with plan IEnumerable<Metric> as per discussion in https://github.com/open-telemetry/opentelemetry-dotnet/pull/2295#discussion_r700611706

3. Modified Console/Prometheus (To demonstrate one push and one pull based). Rest of the Exporters, and clean up of existing MetricProcessor,Push/PullProcessor will be follow up PR to avoid making this PR even bigger. (Same for proper implementation/lack of it for Flush/Shutdown etc)

4. Addressed this comment: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2141#discussion_r670898590

### CODE REVIEW

File: src/OpenTelemetry/Batch.cs [hunk 1]
```diff
@@ -19,7 +19,6 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using OpenTelemetry.Internal;
-using OpenTelemetry.Metrics;
 
 namespace OpenTelemetry
 {
```
Comment (cijothomas) [inline src/OpenTelemetry/Batch.cs:24]: This file is restored to match the main branch.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 2]
```diff
@@ -174,7 +165,20 @@ private Batch<Metric> Collect()
...
+                    // We cannot simply return the internal structure (array)
+                    // as the user is not expected to navigate it.
+                    // properly.
+                    static IEnumerable<Metric> Iterate(Metric[] metrics, long targetCount)
+                    {
+                        for (int i = 0; i < targetCount; i++)
+                        {
+                            // Check if the Metric has valid
+                            // entries and skip, if not.
+                            yield return metrics[i];
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:179]: I'm pretty sure yields always compile into an allocating structure.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 3]
```diff
@@ -33,28 +33,39 @@ public class MeterProviderSdk
...
         private readonly List<MetricProcessor> metricProcessors = new List<MetricProcessor>();
+        private readonly List<MetricReader> metricReaders = new List<MetricReader>();
         private int metricIndex = -1;
 
         internal MeterProviderSdk(
             Resource resource,
             IEnumerable<string> meterSources,
             List<MeterProviderBuilderSdk.InstrumentationFactory> instrumentationFactories,
-            MetricProcessor[] metricProcessors)
+            MetricProcessor[] metricProcessors,
```
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:43]: Looks like MetricProcessor is not used anymore. Is the plan to remove in a follow up?

File: src/OpenTelemetry.Exporter.Console/ConsoleExporterOptions.cs [hunk 4]
```diff
@@ -29,9 +31,9 @@ public class ConsoleExporterOptions
         public int MetricExportIntervalMilliseconds { get; set; } = 1000;
 
         /// <summary>
-        /// Gets or sets a value indicating whether to export Delta
-        /// values or not (Cumulative).
+        /// Gets or sets the AggregationTemporality used for Sum, Histogram
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Console/ConsoleExporterOptions.cs:34]:
```suggestion
        /// Gets or sets the AggregationTemporality used for Histogram and Sum
```

File: src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    public class BaseExportingMetricReader : MetricReader
+    {
+        private MetricExporter exporter;
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs:24]:
```suggestion
        private readonly MetricExporter exporter;
```

File: src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            this.exporter = exporter;
+        }
+
+        public override void OnCollect(IEnumerable<Metric> metrics)
+        {
+            this.exporter.Export(metrics);
+        }
+
+        public override AggregationTemporality GetAggregationTemporality()
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs:36]: Do we want this to be a method or a get property?

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 7]
```diff
@@ -183,5 +187,25 @@ private Batch<Metric> Collect()
...
+            {
+                foreach (var item in this.instrumentations)
+                {
+                    (item as IDisposable)?.Dispose();
+                }
+
+                this.instrumentations.Clear();
+            }
+
+            foreach (var processor in this.metricProcessors)
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:203]: What about readers?

File: src/OpenTelemetry/Metrics/MetricExporter.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public virtual AggregationTemporality GetAggregationTemporality()
+        {
+            return AggregationTemporality.Cumulative;
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/MetricExporter.cs:72]: Do we want this to be `AggregationTemporality.Unspecified`?

File: src/OpenTelemetry/Metrics/PeriodicExportingMetricReader.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        private MetricExporter exporter;
+        private Task exportTask;
+        private CancellationTokenSource token;
+
+        public PeriodicExportingMetricReader(MetricExporter exporter, int exportIntervalMs)
+        {
+            this.exporter = exporter;
+            this.token = new CancellationTokenSource();
+            this.exportTask = new Task(() =>
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/PeriodicExportingMetricReader.cs:34]: Probably add a TODO here, I hope we can make this running on a dedicated thread (so we can use thread runaway time to see the actual serialization/exporting CPU consumption, and CSwitch). The idea is very similar to how Finalizers run in their dedicated threads.

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MetricExporter.cs (hunk 8 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricExporter.cs:72]: For exporter, I would suggest have two GET properties `MetricExporter.SupportedAggregationTemporality` and `MetricExporter.PreferredAggregationTemporality`.

```csharp
[Flags]
public enum AggregationTemporality : byte
{
    Unspecified = 0x00,
    Cumulative = 0x01,
    Delta = 0x02,
    Both = 0x03,
}
```

| Exporter   | SupportedAggregationTemporality | PreferredAggregationTemporality  |
| ----------- | ----------- | ---- |
| Console | Both | Cumulative |
| InMemory | Both | Unspecified |
| OTLP | Both | Unspecified |
| Prometheus | Cumulative | Cumulative |
| Statsd | Delta | Delta |

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:43]: yes a lot of cleanup coming..

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:179]: will come back to this.

File: src/OpenTelemetry/Metrics/MetricExporter.cs (hunk 8 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MetricExporter.cs:72]: I had thought that Cumulative and Delta were required for Prometheus and Statsd, respectively. If my assumption is right then what's the purpose of `PreferredAggregationTemporality`?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MetricExporter.cs:72]: For Prometheus/StatsD, both would be same. "Preferred" is for those exporters capable of handling both.

Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricExporter.cs:72]: Here is my thinking:
* There will be exporters that:
  1. only support Delta, so we have to either convert Cumulative to Delta, or drop the data, or do something else - e.g. treat it as Gauge 2
  2. only support Cumulative, so we have to either convert Delta to Cumulative, or drop the data
  3. supports both Cumulative and Delta (e.g. Console and InMemory exporter), but would prefer a certain way by default, and allow the user to change the behavior

**Pull request merged**